### PR TITLE
[DataGrid] Make rows immutable for better developer experience

### LINKED ISSUE #1553: [DataGrid] Unable to erase a record when using the "getRowId()"
- [x] The issue is present in the latest release.
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯
I can't erase a row from the data-grid, when the data-grid is using the `getRowId`.

It seems most same like with the #607, only whether using `getRowId` or not is different.

image

## Expected Behavior 🤔
Be able to erase a row from the data-grid even using the `getRowId` method.

## Steps to Reproduce 🕹

https://codesandbox.io/s/loving-butterfly-fsqn4?file=/src/CategoryGrid.tsx

  1. Select any item from the data-grid
  2. Click the `Erase` button
  3. The undefined error would be occured

## Context 🔦
I want to erase a row from the data-grid when using the `getRowId()`

## Your Environment 🌎
  - @material-ui/core 4.11.4
  - @material-ui/data-grid. 4.0.0-alpha.27
  - react 17.0.2
  - react-dom 17.0.0
  - react-scripts 4.0.0
  - Same error occurs in every browsers

### PR DESCRIPTION
**Pull request by ZeeshanTamboli**
Resolves #1553 

1 level freeze on the rows prop. 
Also typed `rows` prop with `Readonly` utility so that it shows compile time error for Typescript devs if they try to mutate rows.

### CODE REVIEW

File: packages/grid/_modules_/grid/GridComponent.tsx [hunk 1]
```diff
@@ -55,6 +55,24 @@ import { visibleGridRowCountSelector } from './hooks/features/filter/gridFilterS
...
+      state,
+      loading,
+      className,
+      licenseStatus,
+      nonce,
+      autoHeight,
+    } = props;
+
+    // freeze rows for immutability
+    Object.freeze(rows);
```
[the remark this answers is not in the document]
Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/GridComponent.tsx:74]: Whats the reason? Why? 🤔

[the remark this answers is not in the document]
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/GridComponent.tsx:74]: The freeze is meant to help developers use the API correctly, to have a fast feedback loop when they do something wrong. In production, it might lead to a crash in edge cases. What's the use case for having it in production?

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx [hunk 2]
```diff
@@ -64,6 +64,15 @@ describe('<DataGrid /> - Layout & Warnings', () => {
       expect(rect.width).to.equal(400 - 2);
     });
 
+    it('should freeze the rows', () => {
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid {...baselineProps} />
+        </div>,
+      );
+      expect(baselineProps.rows).to.be.frozen;
```
[the remark this answers is not in the document]
Comment (ZeeshanTamboli) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:73]: Function call is not working. I also tried:
```suggestion
      // eslint-disable-next-line no-unused-expressions
      expect(baselineProps.rows).to.be.frozen;
```
but still eslint fails:
```
1:1  error  Unused eslint-disable directive (no problems were reported from 'no-unused-expressions')
74:7  error  Expected an assignment or function call and instead saw an expression                     @typescript-eslint/no-unused-expressions
```

File: packages/grid/_modules_/grid/GridComponent.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/GridComponent.tsx:74]: Like what if they ignore or miss that in dev mode but still keep mutating in prod build.

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (ZeeshanTamboli) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:73]: Although this is working:
```suggestion
     // eslint-disable-next-line
      expect(baselineProps.rows).to.be.frozen;
```

[the remark this answers is not in the document]
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:73]: What about rewriting the test to be use case driven, meaning asserting that it throws if a developer tries to mutate the rows?

[the remark this answers is not in the document]
Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:73]: Otherwise:

```suggestion
      expect(Object.isFrozen(baselineProps.rows)).to.equal(true);
```

[the remark this answers is not in the document]
Comment (ZeeshanTamboli) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx]: @oliviertassinari I have added a test case in commit https://github.com/mui-org/material-ui-x/pull/1661/commits/396f95d2f3ce7d030e1af2ac2d30c328c608f311 but I am not able to figure out why is this failing and how to test the error that is thrown. If you could lend a hand on this?

[the remark this answers is not in the document]
Comment (ZeeshanTamboli) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx]: I think it's because it is native browser error and not thrown as user-defined exception?

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx [hunk 3]
```diff
@@ -64,6 +68,35 @@ describe('<DataGrid /> - Layout & Warnings', () => {
...
+            </div>
+            <button onClick={addRow}>Add a new row</button>
+          </React.Fragment>
+        );
+      };
+      render(<TestDataGrid />);
+      const addRowButton = screen.getByText('Add a new row');
+      expect(() => {
+        fireEvent.click(addRowButton);
+      }).to.throw();
```
Comment (ZeeshanTamboli) [inline packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:97]: This use-case driven test is still failing. Can't figure out why. Need some direction.

**Comment by ZeeshanTamboli** [comment]
Wrote a test case resembling a user (not passing), otherwise implementation test case below:
```jsx
it('should freeze the rows', () => {
      render(
        <div style={{ width: 300, height: 300 }}>
          <DataGrid {...baselineProps} />
        </div>,
      );
     expect(Object.isFrozen(baselineProps.rows)).to.equal(true);
});
```

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx (hunk 3 above)
Comment (m4theushw) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:97]: The error isn't throw at the click event, but during the `rows.push`. You have to wrap the entire test case in an `expect` to properly catch it. Try this:

```js
  it('should throw an error if rows props is being mutated', () => {
    expect(() => {
      // We don't want to freeze baselineProps.rows
      const rows = [...baselineProps.rows];
      render(
        <div style={{ width: 300, height: 300 }}>
          <DataGrid {...baselineProps} rows={rows} />
        </div>,
      );
      rows.push({ id: 3, brand: 'Louis Vuitton' });
    }).to.throw('Cannot add property 3, object is not extensible');
  });
```

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx [hunk 4]
```diff
@@ -64,6 +68,35 @@ describe('<DataGrid /> - Layout & Warnings', () => {
       expect(rect.width).to.equal(400 - 2);
     });
 
+    it('should throw an error if rows props is being mutated. Error should be thrown due to Object.freeze(rows) on rows prop', () => {
```
Comment (m4theushw) [inline packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:71]: Test cases inside the "Layout" block only run in a real browser with `yarn test:karma`. Since this test doesn't require one, you can move it to outside to be able to run it with `yarn test:unit`.

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx (hunk 3 above)
Comment (ZeeshanTamboli) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx]: Thanks @m4theushw . Also nice catch on `baselineProps.rows` should not be freezed. 👍 

Other thing, I removed the error message string assertion `Cannot add property 3, object is not extensible` as it is different for different browsers.
For Firefox:
```
TypeError: can't define array index property past the end of an array with non-writable length
```

For Safari:
```
TypeError: Attempted to assign to readonly property.
```

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx (hunk 4 above)
Comment (ZeeshanTamboli) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx]: Thanks @m4theushw . I moved it outside.
But I checked that `yarn test:karma` runs tests even if it is outside the `Layout`, `warning` blocks. But `yarn test:unit` runs **only** those tests which are outside those blocks. 
So by moving outside, it is still running within the browser. 

Also curious regarding where is that kind of configuration setup?

File: packages/grid/_modules_/grid/GridComponent.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/GridComponent.tsx:73]: > what if

We take the risk to break the UX for a change that was done for DX motivations.

File: packages/grid/_modules_/grid/GridComponent.tsx [hunk 5]
```diff
@@ -54,6 +54,26 @@ import { visibleGridRowCountSelector } from './hooks/features/filter/gridFilterS
...
+      columns,
+      components: componentsProp,
+      componentsProps,
+      getRowId,
+      licenseStatus,
+      loading,
+      nonce,
+      rows,
+      state,
+    } = props;
```
Comment (dtassone) [inline packages/grid/_modules_/grid/GridComponent.tsx:70]: Not sure it is worth destructuring the props in this component as the goal is to keep it as short as possible and just have a list of hooks together. 
The idea is to easily recompose it in other subsets, so the more verbose the least obvious it becomes.
Can we have a hook to freeze rows, so we keep this file clean.
useGridFreezeRows or something like that.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/GridComponent.tsx:70]: +1 for moving the freeze into the existing `useGridRows` 👍

Comment (dtassone) [reply packages/grid/_modules_/grid/GridComponent.tsx]: Well why not treat it seperately. So it could be recomposed.

Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/GridComponent.tsx]: I added it separately in `useGridFreezeRows` as `rows` is used for sorting and filtering hooks as well -- `useGridSorting`, `useGridFilter`. I was thinking about hooks order. So better separate. Now it looks less verbose. Also removed de-structuring of props.

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/GridComponent.tsx]: I was proposing useGridRows because we likely want to freeze the rows provided with the apiRef. So we could keep the related logic in the same place.

Comment (dtassone) [reply packages/grid/_modules_/grid/GridComponent.tsx]: I prefer separate as well, as one could decide to use a library like immer or immutable js and this might conflict with it so if it is embedded in useGridRows it would be impossible to just remove this part.
Also I would argue that it is a separate concern, just like useGridEditRows...

File: packages/storybook/src/stories/grid-sorting.stories.tsx [hunk 6]
```diff
@@ -435,8 +435,9 @@ function sortServerRows(rows: any[], params: GridSortModelParams): Promise<any[]
       }
       const sortedCol = params.sortModel[0];
       const comparator = params.columns[0].sortComparator!;
+      const clonedRows = [...rows];
       let sortedRows = [
-        ...rows.sort((a, b) => comparator(a[sortedCol.field], b[sortedCol.field], a, b)),
+        ...clonedRows.sort((a, b) => comparator(a[sortedCol.field], b[sortedCol.field], a, b)),
```
Comment (dtassone) [inline packages/storybook/src/stories/grid-sorting.stories.tsx:440]: why double `...clonedRows`  and `const clonedRows = [...rows];` ?

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx [hunk 7]
```diff
@@ -32,6 +32,19 @@ describe('<DataGrid /> - Layout & Warnings', () => {
     columns: [{ field: 'brand' }],
   };
 
+  it('should throw an error if rows props is being mutated', () => {
+    expect(() => {
+      // We don't want to freeze baselineProps.rows
+      const rows = [...baselineProps.rows];
```
Comment (dtassone) [inline packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:38]: it makes me want to add an options prop `disableFreezeRowsInDev`, what do you think? 🤔

Comment (oliviertassinari) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:38]: I don't understand the problem this prop would solve. The line looks expected. `baselineProps` is shared between the tests, we can't mutate it.

File: packages/grid/_modules_/grid/GridComponent.tsx (hunk 5 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/GridComponent.tsx]: Ok. The weird part for me is that `useGridEditRows` implements a feature. On the other hand, `useGridFreezeRows` is only here to avoid a foot gun for the developers, to accommodate the implementation tradeoffs of `useGridRows`, relying on immutability. It seems that `useGridFreezeRows` should always come with `useGridRows` (no valid use case without).

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx (hunk 7 above)
Comment (dtassone) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:38]: Just an idea that came up when I saw that. Maybe users would want to share rows in tests or somehow opt out...
I think we can wait for a new issue if it comes up.

File: packages/grid/_modules_/grid/GridComponent.tsx (hunk 5 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/GridComponent.tsx]: Well `useGridRows` is so core that it will be difficult to do anything without it.

File: packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx (hunk 7 above)
Comment (ZeeshanTamboli) [reply packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx:38]: I'd wait for a new issue to tackle it later.

File: packages/storybook/src/stories/grid-sorting.stories.tsx (hunk 6 above)
Comment (ZeeshanTamboli) [reply packages/storybook/src/stories/grid-sorting.stories.tsx]: Oh yes. I removed double spreading now.

**Review by dtassone — APPROVED** [review]
